### PR TITLE
update for 2018 samples

### PR DESCRIPTION
**Pull request by jalmond**
This is first attempt for changes needed for 2018.

### CODE REVIEW

File: SKFlatMaker/src/SKFlatMaker.cc [hunk 1]
```diff
@@ -2622,15 +2629,68 @@ void SKFlatMaker::fillJet(const edm::Event &iEvent)
...
-      tightJetID = (NEMF>0.02 && NEMF<0.99) && (NumNeutralParticles>2) && abs(eta)>2.7 && abs(eta)<=3.0;
-      tightLepVetoJetID = false; //FIXME default is false or true? following CAT2016
+
+    if(DataYear==2016){
+      //=== 2016 jet IDs
+      //==== https://twiki.cern.ch/twiki/bin/view/CMS/JetID13TeVRun2016
+      
+      if(fabs(eta)>3.0){
+	tightJetID = (NEMF<0.90 && NumNeutralParticles>10 );
+	tightLepVetoJetID = (NEMF<0.90 && NumNeutralParticles>10 );
```
Comment (jedori0228) [inline SKFlatMaker/src/SKFlatMaker.cc]: TightLepVeto is not supported for |eta| > 2.7. Should we put false for this cases?

Comment (jalmond) [reply SKFlatMaker/src/SKFlatMaker.cc]: I updated the PR as suggested

File: SKFlatMaker/src/SKFlatMaker.cc [hunk 2]
```diff
@@ -2917,15 +2969,59 @@ void SKFlatMaker::fillFatJet(const edm::Event &iEvent)
...
-      tightLepVetoJetID = false; //FIXME default is false or true? following CAT2016
+    else   if(DataYear<=2017){
+      //=== 2017 jet IDs                                                                                                                                                                                    
+      //==== https://twiki.cern.ch/twiki/bin/view/CMS/JetID13TeVRun2017                                                                                                                                    
+      if(fabs(eta)>3.0){
+        tightJetID = (NEMF<0.90 && NHF>0.02 && NumNeutralParticles> 2 && NumNeutralParticles < 15);
+        tightLepVetoJetID = (NEMF<0.90 && NHF>0.02 && NumNeutralParticles>2 && NumNeutralParticles < 15);
+      }
+      else if(fabs(eta)>2.7){
+        tightJetID = (NEMF<0.99);
```
Comment (jedori0228) [inline SKFlatMaker/src/SKFlatMaker.cc]: I think this should be Neutral Hadron Fraction not NEMF

File: SKFlatMaker/src/SKFlatMaker.cc [hunk 3]
```diff
@@ -2622,15 +2629,60 @@ void SKFlatMaker::fillJet(const edm::Event &iEvent)
...
+	tightJetID        = (NHF<0.90 && NEMF<0.90 && NumConst>1) &&            ((fabs(eta)<=2.4 && CHF>0 && CHM>0)              || fabs(eta)>2.4) && fabs(eta)<=2.7;
+	tightLepVetoJetID = (NHF<0.90 && NEMF<0.90 && NumConst>1 && MUF<0.8) && ((fabs(eta)<=2.4 && CHF>0 && CHM>0 && CEMF<0.80) || fabs(eta)>2.4) && fabs(eta)<=2.7;	
+      }
     }
-
+    else{
+      //=== 2018 jet ID
+      //==== https://twiki.cern.ch/twiki/bin/viewauth/CMS/JetID13TeVRun2018
+      if(fabs(eta)>3.0){
+	tightJetID = (NEMF<0.90 && NHF>0.2 && NumNeutralParticles>10 );
```
Comment (jedori0228) [inline SKFlatMaker/src/SKFlatMaker.cc]: NHF>0.2 should be NHF>0.02

File: SKFlatMaker/src/SKFlatMaker.cc [hunk 4]
```diff
@@ -2622,15 +2629,60 @@ void SKFlatMaker::fillJet(const edm::Event &iEvent)
...
+      }
+      else {
+	tightJetID        = (NHF<0.90 && NEMF<0.90 && NumConst>1) &&            ((fabs(eta)<=2.4 && CHF>0 && CHM>0 && CEMF<0.99) || fabs(eta)>2.4) && fabs(eta)<=2.7;
+	tightLepVetoJetID = (NHF<0.90 && NEMF<0.90 && NumConst>1 && MUF<0.8) && ((fabs(eta)<=2.4 && CHF>0 && CHM>0 && CEMF<0.90) || fabs(eta)>2.4) && fabs(eta)<=2.7;
+      }
     }
-    else if(fabs(eta)>2.7){
-      tightJetID = (NEMF>0.02 && NEMF<0.99) && (NumNeutralParticles>2) && abs(eta)>2.7 && abs(eta)<=3.0;
-      tightLepVetoJetID = false; //FIXME default is false or true? following CAT2016
+    else   if(DataYear<=2017){
```
Comment (jedori0228) [inline SKFlatMaker/src/SKFlatMaker.cc]: DataYear==2017

File: SKFlatMaker/src/SKFlatMaker.cc [hunk 5]
```diff
@@ -2622,15 +2629,60 @@ void SKFlatMaker::fillJet(const edm::Event &iEvent)
...
+        tightJetID = (NEMF>0.02 && NEMF<0.99) && (NumNeutralParticles>2);                                                                                                                                 
+        tightLepVetoJetID =  (NEMF>0.02 && NEMF<0.99) && (NumNeutralParticles>2);
+      }
+      else{
+	tightJetID        = (NHF<0.90 && NEMF<0.90 && NumConst>1) &&            ((fabs(eta)<=2.4 && CHF>0 && CHM>0)              || fabs(eta)>2.4) && fabs(eta)<=2.7;
+	tightLepVetoJetID = (NHF<0.90 && NEMF<0.90 && NumConst>1 && MUF<0.8) && ((fabs(eta)<=2.4 && CHF>0 && CHM>0 && CEMF<0.80) || fabs(eta)>2.4) && fabs(eta)<=2.7;	
+      }
     }
-
+    else{
```
Comment (jedori0228) [inline SKFlatMaker/src/SKFlatMaker.cc]:
```suggestion
    else if(DataYear==2018){
```
For the safety reason, I think we'd better specify the year and add
`else{`
`    cout << "DataYear is not set : DataYear = " << DataYear << endl;`
`    exit(EXIT_FAILURE);`
`}`
after this statement

File: SKFlatMaker/src/SKFlatMaker.cc [hunk 6]
```diff
@@ -2917,15 +2969,59 @@ void SKFlatMaker::fillFatJet(const edm::Event &iEvent)
...
+      else{
+        tightJetID        = (NHF<0.90 && NEMF<0.90 && NumConst>1) &&            ((fabs(eta)<=2.4 && CHF>0 && CHM>0)              || fabs(eta)>2.4) && fabs(eta)<=2.7;
+        tightLepVetoJetID = (NHF<0.90 && NEMF<0.90 && NumConst>1 && MUF<0.8) && ((fabs(eta)<=2.4 && CHF>0 && CHM>0 && CEMF<0.80) || fabs(eta)>2.4) && fabs(eta)<=2.7;
+      }
+    }
+    else{
+      //=== 2018 jet ID                                                                                                                                                                                     
+      //==== https://twiki.cern.ch/twiki/bin/viewauth/CMS/JetID13TeVRun2018                                                                                                                                 
+      if(fabs(eta)>3.0){
+        tightJetID = (NEMF<0.90 && NHF>0.2 && NumNeutralParticles>10 );
```
Comment (jedori0228) [inline SKFlatMaker/src/SKFlatMaker.cc]: Number of Neutral Particles | >2 and < 15
-- | --

Comment (jedori0228) [inline SKFlatMaker/src/SKFlatMaker.cc]: NHF>0.2 should be NHF>0.02

File: SKFlatMaker/src/SKFlatMaker.cc (hunk 2 above)
Comment (jalmond) [reply SKFlatMaker/src/SKFlatMaker.cc]: Yes I fixed it in new PR.

File: SKFlatMaker/src/SKFlatMaker.cc (hunk 3 above)
Comment (jedori0228) [reply SKFlatMaker/src/SKFlatMaker.cc]: This was actually correct, the table in the twiki just fixed.

File: SKFlatMaker/src/SKFlatMaker.cc (hunk 6 above)
Comment (jedori0228) [reply SKFlatMaker/src/SKFlatMaker.cc]: This was actually correct, the table in the twiki just fixed.

File: SKFlatMaker/src/SKFlatMaker.cc (hunk 4 above)
Comment (jalmond) [reply SKFlatMaker/src/SKFlatMaker.cc]: Ok, I assumed that data year could only be 2016/2017/2018, but i see your point and i made the change.

File: SKFlatMaker/src/SKFlatMaker.cc (hunk 5 above)
Comment (jalmond) [reply SKFlatMaker/src/SKFlatMaker.cc]: Made this change in new PR

File: SKFlatMaker/src/SKFlatMaker.cc (hunk 6 above)
Comment (jalmond) [reply SKFlatMaker/src/SKFlatMaker.cc]: The original values are correct right?

Comment (jedori0228) [reply SKFlatMaker/src/SKFlatMaker.cc]: Yep, 0.2 was the correct value.

**Pull request merged**